038-optimization-proposal-autoapproval.md

```diff
@@ -0,0 +1,99 @@
+# Auto-approval mechanism for optimization proposals
+
+This proposal is about allowing the user to create a `KafkaRebalance` custom resource for getting an optimization proposal and running a rebalance, but without the need of the manual approval via the `strimzi.io/rebalance=approve` annotation.
+It means that the user can just create the `KafkaRebalance` custom resource and the corresponding generated optimization proposal will be automatically approved.
+
+## Current situation
+
+Currently, when the user wants to run a cluster rebalance, he has to create a `KafkaRebalance` custom resource for getting an optimization proposal first.
+After getting the proposal, the only way to start the rebalancing operation is by approving it by annotating the custom resource with the `strimzi.io/rebalance=approve` annotation.
+There is the need for a manual intervention of the user and this is a two step process.
+
+## Motivation
+
+A user would have the desire to reduce the manual intervention on a cluster for rebalancing operations.
+Automatically approving an optimization proposal can save time from an operational point of view.
+It enables more automation where just creating a `KafkaRebalance` custom resource can go straight to the cluster rebalance without any additional manual interaction.
+
+## Proposal
+
+The `KafkaRebalance` custom resource can be extended with a new `spec.autoApproval` field for this purpose.
+The auto-approval can then be enabled or disable through a boolean `spec.autoApproval.enabled` field.
+
+### Auto-approval
+
+The auto-approval can be enabled through the boolean `spec.autoApproval.enabled` field set to `true`.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval:
+    enabled: true
+```
+
+The same result can be easily achieved if the `spec.autoApproval` is specified as empty.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval: {}
+```
+
+The reason for not having a simple boolean `spec.autoApproval` field is for leaving space for a future extension, maybe adding some criteria or rules that have to be matched for auto-approving the optimization proposal.
+
+### No auto-approval (default)
+
+If the `spec.autoApproval` field is not specified at all, the default behavior will be the current one, so the need for manual approval by the user annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
+
+Of course, setting the `spec.autoApproval.enabled` to `false` drives to the same result.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: # any mode
+  autoApproval:
+    enabled: false
+```
```

Comment:
I wonder what others think about this. It might be a bit confusing to the users:
* This defaults to disabled:
  ```yaml
   spec:
     mode: ...
  ```
  => In other words, `enabled` seems to default to false.
* This defaults to enabled:
  ```yaml
   spec:
     mode: ...
     autoApproval: {}
  ```
  => Now suddenly, `enabled` seems to default to true.
* This seems to make not much sense (why would you do this if you can just leave this section out?)
  ```yaml
   spec:
     mode: ...
     autoApproval:
       enabled: false
  ```
* It would be especially weird in the future if you have something like this
  ```yaml
   spec:
     mode: ...
     autoApproval:
       enabled: false
       criteria:
         - ...
         - ...
  ```

I understand the motivation and it can be implemented in the code reasonably. I just wonder how intuitive this will be for a new user who will not read the proposal and know the background.

Would it be more natural to remove the `enabled` field completely and have it just as an empty object? I.e.
* This would disable it
  ```yaml
   spec:
     mode: ...
  ```
* This will enable it:
  ```yaml
   spec:
     mode: ...
     autoApproval: {}
  ```

I know the empty object might seem a bit weird, but if it is an future entry point, why not use it that way? And at the end, most Kafka CRs already use this in a way `userOperator: {}` => enabled with any flag.

Reply:
That's a really good point @scholzj, the behaviour of `enabled` is not unique in this way.
Thinking about having `autoApproval` section with `enabled: false` and maybe some criteria could be a way to "temporary" disable auto-approval? But does it really make any sense? I don't think so.
I will go with your suggestion getting rid of the `enabled` flag even because it could simplify the logic a little bit doesn't taking into account weird situations like the one just described.